engine/uState.ml

```diff
@@ -207,6 +209,12 @@ let undefined m =
   let mq = QMap.filter (fun _ v -> Option.is_empty v) m.qmap in
   QMap.domain mq
 
+let restrict m qvars =
+  let qvars = QSet.union qvars m.rigid in
+  { rigid = m.rigid;
+    qmap = QMap.filter (fun qv _ -> QSet.mem qv qvars) m.qmap;
```

Comment:
possibly the defined qvars should also be kept 
for universes restrict removes undefined univs but keeps the defined ones